apollo-federation/src/query_plan/query_planning_traversal.rs

```diff
@@ -339,6 +354,7 @@ impl<'a: 'b, 'b> QueryPlanningTraversal<'a, 'b> {
     )]
     fn find_best_plan_inner(&mut self) -> Result<Option<&BestQueryPlanInfo>, FederationError> {
         while !self.open_branches.is_empty() {
+            self.parameters.check_cancellation()?;
```

Comment:
Checks like this would need to be added to all loops where the planner can spend significant amount of time.

Reply:
I think `handle_open_branch` is the one. Other functions won't take too long. Even if `advance_with_operation_element` created a million options, it returns reasonably quickly. It's the `handle_open_branch` that actually visits them. One decision point is, `handle_open_branch` is called recursively, so we may decide whether it checks only at the top-level or always.